Unzip task fails when trying to create directories.

### Steps to reproduce

I have the following project file. During the build step many errors are produced that it can't create files. All the files are laid down correctly and the files that it thinks it can't create are actually directories.

Project file
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>netstandard2.0;net46</TargetFrameworks>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
  </PropertyGroup>
  
  <Target Name="DownloadContentFiles">
    <DownloadFile
        SourceUrl="https://github.com/rprichard/winpty/releases/download/0.4.3/winpty-0.4.3-msvc2015.zip"
        DestinationFolder="$(OutputPath)">
      <Output TaskParameter="DownloadedFile" ItemName="Content" />
    </DownloadFile>
  </Target>

  <Target Name="UnzipArchive" BeforeTargets="Build" DependsOnTargets="DownloadContentFiles">
    <Unzip
        SourceFiles="$(OutputPath)\winpty-0.4.3-msvc2015.zip"
        DestinationFolder="$(OutputPath)\unzipped"
        OverwriteReadOnlyFiles="true"
        ContinueOnError="true"
        />
  </Target>
</Project>
```

### Expected  behavior
The build should complete with no errors.

### Actual behavior
I get spammed with errors similar to the following:

```
C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "ia32_xp/bin/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\ia32_xp\bin\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\ia32_xp\bin\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "ia32_xp/lib/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\ia32_xp\lib\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\ia32_xp\lib\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "x64/bin/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\x64\bin\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\x64\bin\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "x64/lib/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\x64\lib\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\x64\lib\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "x64_xp/bin/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\x64_xp\bin\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\x64_xp\bin\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "x64_xp/lib/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\x64_xp\lib\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\net46\unzipped\x64_xp\lib\'.
1>Done building project "Pty.Net.csproj".
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "ia32_xp/bin/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\ia32_xp\bin\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\ia32_xp\bin\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "ia32_xp/lib/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\ia32_xp\lib\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\ia32_xp\lib\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "x64/bin/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\x64\bin\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\x64\bin\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "x64/lib/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\x64\lib\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\x64\lib\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "x64_xp/bin/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\x64_xp\bin\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\x64_xp\bin\'.
1>C:\Users\micro\source\repos\Pty.Net\src\Pty.Net\Pty.Net.csproj(17,5): warning MSB3936: Failed to open unzip file "x64_xp/lib/" to "C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\x64_xp\lib\".  Could not find a part of the path 'C:\Users\micro\source\repos\Pty.Net\bin\Pty.Net\Debug\netstandard2.0\unzipped\x64_xp\lib\'.
```

### Environment data
`msbuild /version` output: 15.8.169.51996

OS info: Windows 10 Version	10.0.17763 Build 17763

If applicable, version of the tool that invokes MSBuild (Visual Studio, dotnet CLI, etc): Visual Studio 15.8

--- a/src/Tasks.UnitTests/Unzip_Tests.cs
+++ b/src/Tasks.UnitTests/Unzip_Tests.cs
@@ -73,6 +73,41 @@ namespace Microsoft.Build.Tasks.UnitTests
             }
         }
 
+        [Fact]
+        public void CanUnzip_ExplicitDirectoryEntries()
+        {
+            using (TestEnvironment testEnvironment = TestEnvironment.Create())
+            {
+                TransientTestFolder source = testEnvironment.CreateFolder(createFolder: true);
+                TransientTestFolder destination = testEnvironment.CreateFolder(createFolder: false);
+                testEnvironment.CreateFile(source, "BE78A17D30144B549D21F71D5C633F7D.txt", "file1");
+                testEnvironment.CreateFile(source, "A04FF4B88DF14860B7C73A8E75A4FB76.txt", "file2");
+                TransientTestFolder emptyDir = source.CreateDirectory("emptyDir");
+                TransientTestFolder subDir = source.CreateDirectory("subDir");
+                subDir.CreateFile("F83E9633685494E53BEF3794EDEEE6A6.txt", "file3");
+                subDir.CreateFile("21D6D4596067723B3AC5DF9A8B3CBFE7.txt", "file4");
+
+                TransientZipArchive zipArchive = TransientZipArchive.Create(source, testEnvironment.CreateFolder(createFolder: true));
+
+                Unzip unzip = new Unzip
+                {
+                    BuildEngine = _mockEngine,
+                    DestinationFolder = new TaskItem(destination.Path),
+                    OverwriteReadOnlyFiles = true,
+                    SkipUnchangedFiles = false,
+                    SourceFiles = new ITaskItem[] { new TaskItem(zipArchive.Path) }
+                };
+
+                unzip.Execute().ShouldBeTrue(() => _mockEngine.Log);
+
+                _mockEngine.Log.ShouldContain(Path.Combine(destination.Path, "BE78A17D30144B549D21F71D5C633F7D.txt"), () => _mockEngine.Log);
+                _mockEngine.Log.ShouldContain(Path.Combine(destination.Path, "A04FF4B88DF14860B7C73A8E75A4FB76.txt"), () => _mockEngine.Log);
+                _mockEngine.Log.ShouldContain(Path.Combine(destination.Path, "subdir", "F83E9633685494E53BEF3794EDEEE6A6.txt"), () => _mockEngine.Log);
+                _mockEngine.Log.ShouldContain(Path.Combine(destination.Path, "subdir", "21D6D4596067723B3AC5DF9A8B3CBFE7.txt"), () => _mockEngine.Log);
+                Directory.Exists(Path.Combine(destination.Path, "emptyDir"));
+            }
+        }
+
         [PlatformSpecific(TestPlatforms.Windows)] // Can't figure out how to make CreateDirectory throw on non-Windows
         [Fact]
         public void LogsErrorIfDirectoryCannotBeCreated()

--- a/src/Tasks/Unzip.cs
+++ b/src/Tasks/Unzip.cs
@@ -131,6 +131,15 @@ namespace Microsoft.Build.Tasks
             {
                 FileInfo destinationPath = new FileInfo(Path.Combine(destinationDirectory.FullName, zipArchiveEntry.FullName));
 
+                // Zip archives can have directory entries listed explicitly.
+                // If this entry is a directory we should create the it and move to the next entry.
+                if (Path.GetFileName(destinationPath.FullName).Length == 0)
+                {
+                    // The entry is a directory
+                    Directory.CreateDirectory(destinationPath.FullName);
+                    continue;
+                }
+
                 if (!destinationPath.FullName.StartsWith(destinationDirectory.FullName, StringComparison.OrdinalIgnoreCase))
                 {
                     // ExtractToDirectory() throws an IOException for this but since we're extracting one file at a time

--- a/src/Tasks/Unzip.cs
+++ b/src/Tasks/Unzip.cs
@@ -132,7 +132,7 @@ namespace Microsoft.Build.Tasks
                 FileInfo destinationPath = new FileInfo(Path.Combine(destinationDirectory.FullName, zipArchiveEntry.FullName));
 
                 // Zip archives can have directory entries listed explicitly.
-                // If this entry is a directory we should create the it and move to the next entry.
+                // If this entry is a directory we should create it and move to the next entry.
                 if (Path.GetFileName(destinationPath.FullName).Length == 0)
                 {
                     // The entry is a directory

